dwave/plugins/torch/samplers/block_spin_update.py

```diff
@@ -0,0 +1,165 @@
+        self.idx_to_node = grbm.idx_to_node
+        self.partition = self.get_partition(topology)
+        self.padded_adjacencies, self.padded_adjacencies_weight = self.get_adjacencies()
+        self.x = nn.Parameter(rands((num_reads, grbm.n_nodes)), requires_grad=False)
+        self.linear = grbm._linear
+        self.quadratic = grbm._quadratic
+
+    @property
+    def kind(self):
+        return self._kind
+
+    def get_adjacencies(self) -> tuple[list[torch.Tensor], list[torch.Tensor]]:
+        """Create two lists of padded adjacency lists (tensors), one for neighbouring indices and
+        another for the corresponding weight indices.
+
+        The issue begins with the adjacency lists being ragged. To address this, we pad adjacencies
+        with -1 values. The exact values do not matter, as the way these adjacencies will be used is
+        by padding an input state with 0s, so when accessing `-1`, the output will be masked out.
+
+        Returns:
+            tuple[list[torch.Tensor], list[torch.Tensor]]: The first output is a padded adjacency
+            list of neighbouring indices where -1 indicates empty. The second output is an adjacency
+            list of weight indices (in sparse indexing) used to retrieve the corresponding edge
+            weight.
+        """
+        max_degree = max([self.G.degree[v] for v in self.G])
+        padded_adjacencies = nn.Parameter(
+            -torch.ones(len(self.G), max_degree, dtype=int), requires_grad=False
+        )
+        padded_adjacencies_weight = nn.Parameter(
+            -torch.ones(len(self.G), max_degree, dtype=int), requires_grad=False
+        )
+
+        adjacency_dict = defaultdict(list)
+        edge_to_idx = dict()
+        for idx, (u, v) in enumerate(
+            zip(self.edge_idx_i.tolist(),
+                self.edge_idx_j.tolist())):
+            adjacency_dict[v].append(u)
+            adjacency_dict[u].append(v)
+            edge_to_idx[u, v] = idx
+            edge_to_idx[v, u] = idx
+        for u in self.idx_to_node:
+            neighbours = adjacency_dict[u]
+            adj_weight_idxs = [edge_to_idx[u, v] for v in neighbours]
+            num_neighbours = len(neighbours)
+            padded_adjacencies[u][:num_neighbours] = torch.tensor(neighbours)
+            padded_adjacencies_weight[u][:num_neighbours] = torch.tensor(adj_weight_idxs)
+        return padded_adjacencies, padded_adjacencies_weight
+
+    def get_partition(self, topology):
+        partition = defaultdict(list)
+        if topology == "zephyr":
+            lin2lattice = dnx.zephyr_coordinates(self.G.graph['rows']).linear_to_zephyr
+            crayon = dnx.zephyr_four_color
+        elif topology == "pegasus":
+            lin2lattice = dnx.pegasus_coordinates(self.G.graph['rows']).linear_to_pegasus
+            crayon = dnx.pegasus_four_color
+        elif topology == "chimera":
+            lin2lattice = dnx.chimera_coordinates(self.G.graph['rows']).linear_to_chimera
```

Comment:
chimera graphs support 2 node types, this is not required if the graph is already coordinated.